File: back-kotlin/src/main/kotlin/me/nicolas/stravastats/domain/business/AnnualGoals.kt
package me.nicolas.stravastats.domain.business

enum class AnnualGoalMetric {
    DISTANCE_KM,
    ELEVATION_METERS,
    MOVING_TIME_SECONDS,
    ACTIVITIES,
    ACTIVE_DAYS,
    EDDINGTON,
}

enum class AnnualGoalStatus {
    NOT_SET,
    AHEAD,
    ON_TRACK,
    BEHIND,
}

data class AnnualGoalTargets(
    val distanceKm: Double? = null,
    val elevationMeters: Int? = null,
    val movingTimeSeconds: Int? = null,
    val activities: Int? = null,
    val activeDays: Int? = null,
    val eddington: Int? = null,
)

data class AnnualGoalProgress(
    val metric: AnnualGoalMetric,
    val label: String,
    val unit: String,
    val current: Double,
    val target: Double,
    val progressPercent: Double,
    val expectedProgressPercent: Double,
    val projectedEndOfYear: Double,
    val requiredPace: Double,
    val requiredPaceUnit: String,
    val status: AnnualGoalStatus,
)

data class AnnualGoals(
    val year: Int,
    val activityTypeKey: String,
    val targets: AnnualGoalTargets,
    val progress: List<AnnualGoalProgress>,
)

File: back-go/internal/shared/domain/business/annualgoals.go
package business

type AnnualGoalMetric string

const (
	AnnualGoalMetricDistanceKm      AnnualGoalMetric = "DISTANCE_KM"
	AnnualGoalMetricElevationMeters AnnualGoalMetric = "ELEVATION_METERS"
	AnnualGoalMetricActivities      AnnualGoalMetric = "ACTIVITIES"
	AnnualGoalMetricActiveDays      AnnualGoalMetric = "ACTIVE_DAYS"
	AnnualGoalMetricEddington       AnnualGoalMetric = "EDDINGTON"
)

type AnnualGoalStatus string

const (
	AnnualGoalStatusNotSet  AnnualGoalStatus = "NOT_SET"
	AnnualGoalStatusAhead   AnnualGoalStatus = "AHEAD"
	AnnualGoalStatusOnTrack AnnualGoalStatus = "ON_TRACK"
	AnnualGoalStatusBehind  AnnualGoalStatus = "BEHIND"
)

type AnnualGoalTargets struct {
	DistanceKm        *float64 `json:"distanceKm,omitempty"`
	ElevationMeters   *int     `json:"elevationMeters,omitempty"`
	MovingTimeSeconds *int     `json:"movingTimeSeconds,omitempty"`
	Activities        *int     `json:"activities,omitempty"`
	ActiveDays        *int     `json:"activeDays,omitempty"`
	Eddington         *int     `json:"eddington,omitempty"`
}

type AnnualGoalProgress struct {
	Metric                  AnnualGoalMetric  `json:"metric"`
	Label                   string            `json:"label"`
	Unit                    string            `json:"unit"`
	Current                 float64           `json:"current"`
	Target                  float64           `json:"target"`
	ProgressPercent         float64           `json:"progressPercent"`
	ExpectedProgressPercent float64           `json:"expectedProgressPercent"`
	ProjectedEndOfYear      float64           `json:"projectedEndOfYear"`
	RequiredPace            float64           `json:"requiredPace"`
	RequiredPaceUnit        string            `json:"requiredPaceUnit"`
	RequiredWeeklyPace      float64           `json:"requiredWeeklyPace"`
	Last30Days              float64           `json:"last30Days"`
	Last30DaysWeeklyPace    float64           `json:"last30DaysWeeklyPace"`
	WeeklyPaceGap           float64           `json:"weeklyPaceGap"`
	SuggestedTarget         *float64          `json:"suggestedTarget,omitempty"`
	Monthly                 []AnnualGoalMonth `json:"monthly"`
	Status                  AnnualGoalStatus  `json:"status"`
}

type AnnualGoalMonth struct {
	Month              int     `json:"month"`
	Value              float64 `json:"value"`
	Cumulative         float64 `json:"cumulative"`
	ExpectedCumulative float64 `json:"expectedCumulative"`
}

type AnnualGoals struct {
	Year            int                  `json:"year"`
	ActivityTypeKey string               `json:"activityTypeKey"`
	Targets         AnnualGoalTargets    `json:"targets"`
	Progress        []AnnualGoalProgress `json:"progress"`
}
